src/mobu/services/github_ci/ci_manager.py

```diff
@@ -0,0 +1,337 @@
+
+from __future__ import annotations
+
+import asyncio
+from asyncio import Queue
+from dataclasses import dataclass, field
+from typing import TypeAlias
+
+from aiojobs import Job, Scheduler
+from httpx import AsyncClient
+from safir.github import GitHubAppClientFactory
+from structlog.stdlib import BoundLogger
+
+from ...config import config
+from ...models.ci_manager import CiManagerSummary, CiWorkerSummary
+from ...models.user import User
+from ...storage.gafaelfawr import GafaelfawrStorage
+from ...storage.github import GitHubStorage
+from .ci_notebook_job import CiNotebookJob
+
+__all__ = ["CiManager"]
+
+
+@dataclass
+class CiManagerLifecycle:
+    marked_remaining: asyncio.Event = field(default_factory=asyncio.Event)
+
+
+@dataclass
+class JobLifecycle:
+    processing: asyncio.Event = field(default_factory=asyncio.Event)
+    processed: asyncio.Event = field(default_factory=asyncio.Event)
+
+
+@dataclass
+class QueuedJob:
+    job: CiNotebookJob
+    lifecycle: JobLifecycle
+
+
+QueueItem: TypeAlias = QueuedJob | None
+
+
+class CiManager:
+    """Manages processing work for GitHub CI checks.
+
+    This should be a process singleton. It is responsible for:
+    * Creating background workers to process GitHub CI events
+    * Ensuring they run at the appropriate level of concurrency given the
+      number of available users
+    * Ensuring GitHub CI checks are not left in a forever-in-progress state
+      when mobu shuts down
+
+    Parameters
+    ----------
+    users
+        A list of static users that are available to run jobs. Each of these
+        users will get assigned to a worker, and will process one job at a
+        time.
+    http_client
+        Shared HTTP client.
+    gafaelfawr_storage
+        Gafaelfawr storage client.
+    logger
+        Global logger to use for process-wide (not monkey) logging.
+    """
+
+    shutdown_error_msg = "Mobu stopped, try re-running this check."
+
+    def __init__(
+        self,
+        users: list[User],
+        http_client: AsyncClient,
+        gafaelfawr_storage: GafaelfawrStorage,
+        logger: BoundLogger,
+    ) -> None:
+        self._users = users
+        self._gafaelfawr = gafaelfawr_storage
+        self._http_client = http_client
+        self._logger = logger.bind(ci_manager=True)
+        self._scheduler: Scheduler = Scheduler()
+        self._queue: Queue[QueueItem] = Queue()
+        self._jobs: list[Job] = []
+        self.workers: list[Worker] = []
+
+        # Used for deterministic testing
+        self.lifecycle = CiManagerLifecycle()
+
+        if not config.github_ci_app.id:
+            raise RuntimeError("MOBU_GITHUB_CI_APP_ID was not set")
+        if not config.github_ci_app.webhook_secret:
+            raise RuntimeError("MOBU_GITHUB_CI_APP_WEBHOOK_SECRET was not set")
+        if not config.github_ci_app.private_key:
+            raise RuntimeError("MOBU_GITHUB_CI_APP_PRIVATE_KEY was not set")
```

Comment:
Here too (future work) this really should be a validator on the model. For Pydantic-based applications, it's usually a bit of code smell if you find yourself doing sanity checks on the model in code outside of the model. Pydantic should be able to give you a validated model. (In this case, you probably need two models, a disabled one and an enabled one, and then all of the underlying code is only created if the thing is enabled and takes the enabled model as an argument. There's an example of this pattern in `controller/src/controller/config.py` in the nublado repository if I remember correctly.

Reply:
Simplified (a bit differently than the disabled/enabled union) in https://github.com/lsst-sqre/mobu/pull/354